x-pack/plugins/global_search_bar/public/components/search_bar.tsx

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import {
+  EuiBadge,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiSelectableTemplateSitewide,
+  EuiSelectableTemplateSitewideOption,
+  EuiText,
+  EuiSelectableMessage,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage } from '@kbn/i18n/react';
+import { ApplicationStart } from 'kibana/public';
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { useDebounce, useEvent } from 'react-use';
+import { GlobalSearchPluginStart, GlobalSearchResult } from '../../../global_search/public';
+
+const useIfMounted = () => {
+  const isMounted = useRef(true);
+  useEffect(
+    () => () => {
+      isMounted.current = false;
+    },
+    []
+  );
+
+  const ifMounted = useCallback((func) => {
+    if (isMounted.current && func) {
+      func();
+    }
+  }, []);
+
+  return ifMounted;
+};
+
+interface Props {
+  globalSearch: GlobalSearchPluginStart['find'];
+  navigateToUrl: ApplicationStart['navigateToUrl'];
+}
+
+const clearField = (field: HTMLInputElement) => {
+  const nativeInputValue = Object.getOwnPropertyDescriptor(HTMLInputElement.prototype, 'value');
+  const nativeInputValueSetter = nativeInputValue ? nativeInputValue.set : undefined;
+  if (nativeInputValueSetter) {
+    nativeInputValueSetter.call(field, '');
+  }
+
+  field.dispatchEvent(new Event('change'));
+};
+
+const cleanMeta = (str: string) => (str.charAt(0).toUpperCase() + str.slice(1)).replace(/-/g, ' ');
+const blurEvent = new FocusEvent('blur');
+
+export function SearchBar({ globalSearch, navigateToUrl }: Props) {
+  const ifMounted = useIfMounted();
+  const [searchValue, setSearchValue] = useState<string>('');
+  const [searchRef, setSearchRef] = useState<HTMLInputElement | null>(null);
+  const [options, _setOptions] = useState([] as EuiSelectableTemplateSitewideOption[]);
+  const isWindows = navigator.platform.toLowerCase().indexOf('win') >= 0;
+
+  const setOptions = useCallback(
+    (_options: GlobalSearchResult[]) => {
+      ifMounted(() =>
+        _setOptions([
+          ..._options.map((option) => ({
+            key: option.id,
+            label: option.title,
+            url: option.url,
+            ...(option.icon && { icon: { type: option.icon } }),
+            ...(option.type &&
+              option.type !== 'application' && { meta: [{ text: cleanMeta(option.type) }] }),
+          })),
+        ])
+      );
+    },
+    [ifMounted, _setOptions]
+  );
+
+  useDebounce(
+    () => {
+      let arr: GlobalSearchResult[] = [];
+      globalSearch(searchValue, {}).subscribe({
+        next: ({ results }) => {
+          if (searchValue.length > 0) {
```

Comment:
If we extracted all of the logic in this callback into a separate function, it'd be much easier to write tests against.

Reply:
Leaving this one open. If others agree with it – please chime in!

The tests right now test everything this function does through using the UI. Extracting the function could make it more easily unit testable but I think doing it all through the UI makes the actual function a implementation detail that we don't need to worry about.

If others disagree, would love to hear it!